### PR TITLE
Fix outlined button stroke defaults

### LINKED ISSUE #18170: ContainedButton has incorrect default border stroke color and size

### PR DESCRIPTION
**Pull request by clocksmith**
Primarily, this change corrects the stroke defaults to color: black 12% alpha and width: 1dp

This change also:
1. Extracts the resolver for outline color into its own function. This type of logic will most likely be consolidated as we improve the theming library.
2. Adds unit tests for the newly affected disabled state, and re-organizes the existing test so that it clearly tests each state for the component specific overrides.
3. Leaves various TODOs for properly honoring the buttonThemeData and themeData constants that do not yet exist.

* I think the dartfmt was correct in these instances, but I apologize in advance if it did the wrong thing.

closes #18170

### CODE REVIEW

File: packages/flutter/lib/src/material/outline_button.dart [hunk 1]
```diff
@@ -354,22 +354,31 @@ class _OutlineButtonState extends State<OutlineButton> with SingleTickerProvider
...
+    if (_pressed && widget.highlightedBorderColor != null) {
+      return widget.highlightedBorderColor;
+    } else if (!widget.enabled && widget.disabledBorderColor != null) {
+      return widget.disabledBorderColor;
+    } else {
+      // TODO(clocksmith): Check theme.surfaceColor once its available.
+      // TODO(clocksmith): Check buttonThemeData once its passed.
+      final Color themeColor = theme.primaryColor ??
+          theme.brightness == Brightness.dark ? Colors.white : Colors.black;
+      return widget.borderSide?.color ?? themeColor.withOpacity(0.12);
```
[the remark this answers is not in the document]
Comment (clocksmith) [reply packages/flutter/lib/src/material/outline_button.dart]: Done

**Comment by clocksmith** [comment]
This is good to go (minus the current flutter-build being broken)

File: packages/flutter/lib/src/material/outline_button.dart [hunk 2]
```diff
@@ -349,27 +351,37 @@ class _OutlineButtonState extends State<OutlineButton> with SingleTickerProvider
...
+  Color _getOutlineColor(ThemeData theme) {
+    if (_pressed && widget.highlightedBorderColor != null) {
+      return widget.highlightedBorderColor;
+    } else if (!widget.enabled && widget.disabledBorderColor != null) {
+      return widget.disabledBorderColor;
+    } else {
+      // TODO(clocksmith): Check theme.surfaceColor once its available.
+      // TODO(clocksmith): Check buttonThemeData once its passed.
+      final Color themeColor = theme.primaryColor ??
+          (theme.brightness == Brightness.dark ? Colors.white : Colors.black);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/outline_button.dart]: use a switch here (see Flutter style guide)

File: packages/flutter/test/material/outline_button_test.dart [hunk 3]
```diff
@@ -26,9 +27,12 @@ void main() {
...
 
     await tester.pumpWidget(
-      buildFrame(() { pressedCount += 1; }),
+      buildFrame(() {
+        pressedCount += 1;
+      }),
     );
-    expect(tester.widget<OutlineButton>(find.byType(OutlineButton)).enabled, true);
+    expect(
+        tester.widget<OutlineButton>(find.byType(OutlineButton)).enabled, true);
```
Comment (Hixie) [inline packages/flutter/test/material/outline_button_test.dart]: please don't indent like this; see style guide

File: packages/flutter/test/material/outline_button_test.dart [hunk 4]
```diff
@@ -223,26 +260,31 @@ void main() {
...
+        tester.getSize(find.byType(Text)).width, isIn(<double>[126.0, 127.0]));
     expect(tester.getSize(find.byType(Text)).height, equals(42.0));
   });
 
-  testWidgets('OutlineButton implements debugFillDescription', (WidgetTester tester) async {
-    final DiagnosticPropertiesBuilder builder = new DiagnosticPropertiesBuilder();
+  testWidgets('OutlineButton implements debugFillDescription',
+      (WidgetTester tester) async {
+    final DiagnosticPropertiesBuilder builder =
+        new DiagnosticPropertiesBuilder();
```
Comment (Hixie) [inline packages/flutter/test/material/outline_button_test.dart]: these are more readable on one line like before

File: packages/flutter/test/material/outline_button_test.dart [hunk 5]
```diff
@@ -43,121 +47,152 @@ void main() {
...
+              )
             ],
-          )
-        ],
-      ),
-      ignoreId: true,
-    ));
+          ),
+          ignoreId: true,
+        ));
```
Comment (Hixie) [inline packages/flutter/test/material/outline_button_test.dart]: why this change? it doesn't seem more readable this way.

maybe dartfmt got confused because there's no trailing commas?

**Review by Hixie — COMMENTED** [review]
if you want to use dartfmt, please increase the line width to avoid egregious line breaking, and be vigilant about trailing commas. 0

**Comment by Hixie** [comment]
i don't really understand #18170; incorrect according to what?

cc @HansMuller

File: packages/flutter/lib/src/material/outline_button.dart (hunk 2 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/outline_button.dart]: done

File: packages/flutter/test/material/outline_button_test.dart (hunk 3 above)
Comment (clocksmith) [reply packages/flutter/test/material/outline_button_test.dart]: done

File: packages/flutter/test/material/outline_button_test.dart (hunk 4 above)
Comment (clocksmith) [reply packages/flutter/test/material/outline_button_test.dart]: done

File: packages/flutter/test/material/outline_button_test.dart (hunk 5 above)
Comment (clocksmith) [reply packages/flutter/test/material/outline_button_test.dart]: I've removed dartfmt

**Comment by clocksmith** [comment]
Comments are addressed, PTAL

**Comment by Hixie** [comment]
cc @HansMuller for review since he's back.

File: packages/flutter/test/material/outline_button_test.dart [hunk 6]
```diff
@@ -43,90 +43,109 @@ void main() {
...
       new Directionality(
         textDirection: TextDirection.ltr,
         child: new Material(
           child: new Center(
-            child: new OutlineButton(
-              onPressed: () { },
-              child: const Text('ABC')
-            ),
+            child:
+                new OutlineButton(onPressed: () {}, child: const Text('ABC')),
```
Comment (willlarche) [inline packages/flutter/test/material/outline_button_test.dart]: Still need `new` and `const` here in Dart 2?

File: packages/flutter/test/material/outline_button_test.dart [hunk 7]
```diff
@@ -166,7 +184,7 @@ void main() {
             data: const MediaQueryData(textScaleFactor: 1.0),
             child: new Center(
               child: new OutlineButton(
-                onPressed: () { },
+                onPressed: () {},
```
Comment (willlarche) [inline packages/flutter/test/material/outline_button_test.dart]: Probably best to put all these style corrections in their own PR. NBD.

**Comment by HansMuller** [comment]
Just FTR the goal here is to fix OutlineButton per https://github.com/flutter/flutter/issues/18170#issuecomment-402230741 by:
- Changing the outline border's width from 2.0 to 1.0
- Changing the outline border's color from Colors.grey[200] to Colors.black.withOpacity(0.12)

Current version:
width2

New version:
width1

I assume that the outline's border width is 1.0 per the little "1" in the [material spec](https://material.io/design/components/buttons.html#specs):

spec

The spec doesn't appear to define the outline's color?

File: packages/flutter/test/material/outline_button_test.dart (hunk 6 above)
Comment (Hixie) [reply packages/flutter/test/material/outline_button_test.dart]: Yes please explicitly specify all news and consts still.
We are working on having a lint that will help us consistently strip them out but we want the codebase to be consistent in the meantime.

**Comment by clocksmith** [comment]
The color can be seen if you click the button on bottom right of the spec
window, then select color in the right hand navigation:

It is #0000001e or r: 0 b: 0 g: 0 a:0.12

#0000001e

R0 G0 B0 A0.12

**Comment by HansMuller** [comment]
@clocksmith Right, the spec says that the color is #0000001e. 

However we believe that it's really the (TBD) onSurfaceColor with 12% opacity.

Do we think the outline is actually supposed to be translucent, or is there another level of indirection here: should the outline color actually be an opaque color created by composing #0000001e with the default background color?

**Comment by HansMuller** [comment]
The existing implementation, which seems to have been rewritten more than was needed, effectively defines 6 values:

```
configuration            widget override or null          current default
---------------------------------------------------------------------------
enabled pressed dark     widget.highlightedBorderColor    theme.primaryColor
enabled pressed light    widget.highlightedBorderColor    theme.primaryColor
enabled dark             widget.borderSide.color          Colors.grey[600]
enabled light            widget.borderSide.color          Colors.grey[200]
disabled dark            widget.disabledBorderColor       Colors.grey[800]
disabled light           widget.disabledBorderColor       Colors.grey[100]
```

Before updating OutlineButton we should understand exactly what the values in the default column should be and be able to point to where they're defined in the spec.
